src/endless/Analytics.cpp

```diff
@@ -0,0 +1,201 @@
+#include "stdafx.h"
+#include "Analytics.h"
+#include <afxinet.h>
+
+extern "C" {
+	#include "rufus.h"
+}
+
+#include "Version.h"
+#include "GeneralCode.h"
+
+#define TRACKING_ID "UA-61784217-3"
+#define REG_SECTION "Analytics"
+#define REG_KEY_CLIENT_ID "ClientID"
+#define REG_KEY_DISABLED "Disabled"
+
+#define SERVER_NAME "www.google-analytics.com"
+#define SERVER_PORT 80
+
+static UINT threadSendRequest(LPVOID pParam)
+{
+	FUNCTION_ENTER;
+
+	CInternetSession session(_T("Endless USB Creator"));
+	MSG msg;
+
+	while (GetMessage(&msg, NULL, WM_APP, WM_APP+1)) {
+		CString *pBody = (CString *)msg.wParam;
+		CStringA bodyUtf8 = CW2A(*pBody);
+
+		try {
+			CString headers = _T("Content-type: application/x-www-form-urlencoded");
+			CHttpConnection *conn = session.GetHttpConnection(_T(SERVER_NAME), (INTERNET_PORT)SERVER_PORT);
+			CHttpFile *file = conn->OpenRequest(CHttpConnection::HTTP_VERB_POST, _T("collect"));
+			if (file) {
+				file->SendRequest(headers, (LPVOID)(LPCSTR)bodyUtf8, bodyUtf8.GetLength());
+				uprintf("Analytics req: %s\n", (LPCSTR)bodyUtf8);
+				delete file;
+			}
+			else {
+				uprintf("Analytics req failed\n");
+			}
+			delete conn;
+		}
+		catch (CInternetException *ex) {
+			TCHAR err[256];
+			ex->GetErrorMessage(err, 256);
+			uprintf("Analytics req error: %ls\n", (LPCTSTR)err);
+		}
+
+		delete pBody;
+
+		if (msg.message == WM_APP+1)
+			break;
+	}
+
+	return 0;
+}
+
+
+Analytics::Analytics()
+{
+	m_disabled = disabled();
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	m_trackingId = CString(_T(TRACKING_ID));
+	loadUuid(m_clientId);
+	m_workerThread = AfxBeginThread(threadSendRequest, NULL);
+}
+
+Analytics::~Analytics()
+{
+}
+
+Analytics *Analytics::instance()
+{
+	static Analytics instance;
+	return &instance;
+}
+
+void Analytics::sessionControl(BOOL start)
+{
+	if (m_disabled) return;
+	FUNCTION_ENTER;
+	CString body;
+	prefixId(body);
+	if (start) {
+		body = body + _T("t=screenview&cd=FirstPage&sc=start");
```

Comment:
Is the t=screenview necessary in addition to the sc=start? Can we not just send the sc events by themselves? This might be why the sessions are not ending cleanly.